### PR TITLE
argparse.SUPPRESS-like functionality for advanced uses

### PR DESCRIPTION
**Pull request by idoby**

### CODE REVIEW

**Comment by lebrice** [comment]
Interesting!
Can you give me an example use-case for this? And how is this different than just parsing the class and then using `dataclasses.asdict` on the result?

**Comment by idoby** [comment]
The use case is somewhat esoteric, I admit, but this is a feature that is available in argparse (and therefore SP) add_argument, so now it can be used with add_arguments as well.

I'm combining configs from multiple sources, so I need to know exactly what was specified on the command line, since the command line is neither the top priority or lowest priority source I have.
This way I can enable specifying each part of the dataclass from a different source and assemble the DC at the end of the entire process.

It sounds complicated but it makes keeping track of my ML experiments super easy.

**Review by lebrice — APPROVED** [review]
Looks fine to me. Let's see if all the tests pass.

Thanks for the PR!

**Comment by idoby** [comment]
(It also affects whether the config MUST be specified on the cmdline, help text etc)

**Comment by idoby** [comment]
I want to test it a little bit more with nested DCs before merging this

**Comment by idoby** [comment]
OK seems to work fine. @lebrice I wrote a small hack when dealing with the boolean test cases there, is there a better way to accomplish this?

File: test/test_default_args.py [hunk 1]
```diff
@@ -81,3 +82,103 @@ class SomeClass:
...
+
+    @dataclass
+    class SomeClass:
+        a: some_type
+        b: int = 5
+
+    parser = ArgumentParser()
+    parser.add_arguments(SomeClass, dest='some_class', default=argparse.SUPPRESS)
+    if some_type == bool:
+        def strbool(s):
```
Comment (lebrice) [inline test/test_default_args.py:124]: Use str2bool, wherever it's defined (probably in utils.py)

File: test/test_default_args.py [hunk 2]
```diff
@@ -81,3 +82,103 @@ class SomeClass:
...
+            if s.lower() in ['false', 'no', 'f']:
+                return False
+            else:
+                return True
+
+        parser.add_argument('--c', type=strbool, default=argparse.SUPPRESS)
+    else:
+        parser.add_argument('--c', type=some_type, default=argparse.SUPPRESS)
```
Comment (lebrice) [inline test/test_default_args.py:132]: Can you mix suppressed and non-suppressed arguments? What happens in that case?

**Review by lebrice — APPROVED** [review]
Seems fine so far, and since it doesn't make any test fail, I don't have any objections to adding this.

Thanks @idoby !

**Comment by idoby** [comment]
@lebrice I'm pretty happy with this now, I'd appreciate your review.

File: test/test_default_args.py (hunk 2 above)
Comment (idoby) [reply test/test_default_args.py:123]: I've added another test case to demonstrate this

**Review by lebrice — APPROVED** [review]
Looks great! One last thing I think might be worthwhile to test would be to add a suppressed dataclass using `add_arguments`, as well as another dataclass with a non-suppressed default using `add_arguments`.

I don't think it's totally necessary for this to be merged, but if you find the time, I'd also appreciate a test for conflicts between two such argument groups (i.e. doing `parser.add_arguments(SomeClass, "foo", default=argparse.SUPPRESS)`, followed by `parser.add_arguments(SomeClass, "bar")` .

I believe this should bring out some bugs. Then again, you don't _have_ to add this test, but if you're willing to do it, I'd appreciate it.

Thanks again :)

**Comment by idoby** [comment]
> add a suppressed dataclass using `add_arguments`, as well as another dataclass with a non-suppressed default using `add_arguments`.

> a test for conflicts between two such argument groups (i.e. doing `parser.add_arguments(SomeClass, "foo", default=argparse.SUPPRESS)`, followed by `parser.add_arguments(SomeClass, "bar")` .

Rebased onto master and added these tests. No bugs seem to have emerged this time.

**Comment by idoby** [comment]
We can have a zoom call and I'll show you :)

**Pull request merged**